data/field_type_test.go

```diff
@@ -54,3 +55,58 @@ func TestFieldTypeConversion(t *testing.T) {
 
 	require.Equal(t, obj.FieldType, objCopy.FieldType)
 }
+
+func TestFieldTypeFor(t *testing.T) {
+	tests := []struct {
+		item interface{}
+		want data.FieldType
+	}{
+		// non null values
+		{item: int8(123), want: data.FieldTypeInt8},
+		{item: int16(123), want: data.FieldTypeInt16},
+		{item: int32(123), want: data.FieldTypeInt32},
+		{item: int64(123), want: data.FieldTypeInt64},
+		{item: uint8(123), want: data.FieldTypeUint8},
+		{item: uint16(123), want: data.FieldTypeUint16},
+		{item: uint32(123), want: data.FieldTypeUint32},
+		{item: uint64(123), want: data.FieldTypeUint64},
+		{item: float32(123), want: data.FieldTypeFloat32},
+		{item: float64(123), want: data.FieldTypeFloat64},
+		{item: true, want: data.FieldTypeBool},
+		{item: "foo", want: data.FieldTypeString},
+		{item: time.Unix(1, 1), want: data.FieldTypeTime},
+		{item: json.RawMessage(`{ "foo" : "bar" }`), want: data.FieldTypeJSON},
+		{item: data.EnumItemIndex(123), want: data.FieldTypeEnum},
+		// nullable values
+		{item: pointer(int8(123)), want: data.FieldTypeNullableInt8},
+		{item: pointer(int16(123)), want: data.FieldTypeNullableInt16},
+		{item: pointer(int32(123)), want: data.FieldTypeNullableInt32},
+		{item: pointer(int64(123)), want: data.FieldTypeNullableInt64},
+		{item: pointer(uint8(123)), want: data.FieldTypeNullableUint8},
+		{item: pointer(uint16(123)), want: data.FieldTypeNullableUint16},
+		{item: pointer(uint32(123)), want: data.FieldTypeNullableUint32},
+		{item: pointer(uint64(123)), want: data.FieldTypeNullableUint64},
+		{item: pointer(float32(123)), want: data.FieldTypeNullableFloat32},
+		{item: pointer(float64(123)), want: data.FieldTypeNullableFloat64},
+		{item: pointer(true), want: data.FieldTypeNullableBool},
+		{item: pointer("foo"), want: data.FieldTypeNullableString},
+		{item: pointer(time.Unix(1, 1)), want: data.FieldTypeNullableTime},
+		{item: pointer(json.RawMessage(`{ "foo" : "bar" }`)), want: data.FieldTypeNullableJSON},
+		{item: pointer(data.EnumItemIndex(123)), want: data.FieldTypeNullableEnum},
+		// untyped values
+		{item: nil, want: data.FieldTypeUnknown},
+		{item: 123, want: data.FieldTypeUnknown},
+		{item: pointer(123), want: data.FieldTypeUnknown},
+		{item: 123.456, want: data.FieldTypeFloat64},
+		{item: pointer(123.456), want: data.FieldTypeNullableFloat64},
+	}
+	for _, tt := range tests {
+		t.Run(tt.want.ItemTypeString(), func(t *testing.T) {
+			require.Equal(t, tt.want, data.FieldTypeFor(tt.item))
+		})
+	}
+}
+
+func pointer[T any](input T) any {
```

Comment:
Should it return `*T` ? That is what it currently is in this https://github.com/grafana/grafana/pull/63900 

```
func Pointer[T any](v T) *T { return &v }
```